### PR TITLE
chore: support range requests in trace viewer http server

### LINKED ISSUE #10333: [BUG] cannot select playback time on html report video
**Context:**
- Playwright Version: 1.16.3
- Operating System: Mac
- Node.js version: 17.0.1
- Browser: Chromium

**Code Snippet**

```typescript
import { test } from "@playwright/test";

test("example", async ({ page }) => {
	await page.goto("https://www.google.com");
	const s = page.locator("input[name=q]");
	await s.fill("example");
	await page.waitForTimeout(5000);
	await s.fill("overwritten");
	await page.waitForTimeout(5000);
});

```

**Describe the bug**

When trying to select other times (rewind/forward) on the video, the clip continues to play from its current location.

When I play the video in `test-results`, I can rewind/forward it as expected

https://user-images.githubusercontent.com/20273647/141875252-7d8f47fe-f6d7-474b-83fb-fa12174f9dae.mp4

### PR DESCRIPTION
**Pull request by mxschmitt**
Fixes #10333

### CODE REVIEW

**Comment by JoelEinbinder** [comment]
Is it feasible for us to write a test for this? Code looks fine but quite tricky.

File: packages/playwright-core/src/utils/httpServer.ts [hunk 1]
```diff
@@ -80,22 +80,69 @@ export class HttpServer {
...
+    response.setHeader('Content-Type', contentType);
+    response.setHeader('Content-Length', content.byteLength);
+    response.end(content);
+  }
+
+  _serveRangeFile(request: http.IncomingMessage, response: http.ServerResponse, absoluteFilePath: string) {
+    const range = request.headers.range!;
+    const size = fs.statSync(absoluteFilePath).size;
+    // Parse the range header: https://datatracker.ietf.org/doc/html/rfc7233#section-2.1
+    const [startStr, endStr] = range.replace(/bytes=/, '').split('-');
```
Comment (dgozman) [inline packages/playwright-core/src/utils/httpServer.ts:110]: There could be multiple ranges in the header, separated by commas. We should either  support it, or error out.

File: packages/playwright-core/src/utils/httpServer.ts [hunk 2]
```diff
@@ -80,22 +80,69 @@ export class HttpServer {
...
+    let start = parseInt(startStr, 10);
```
Comment (dgozman) [inline packages/playwright-core/src/utils/httpServer.ts:111]: `parseInt` is happy to parse "10abc". Use `+startStr` and check for NaN instead.

File: packages/playwright-core/src/utils/httpServer.ts [hunk 3]
```diff
@@ -80,22 +80,69 @@ export class HttpServer {
...
+
+    // Sending Partial Content: https://datatracker.ietf.org/doc/html/rfc7233#section-4.1
+    response.writeHead(206, {
+      'Content-Range': `bytes ${start}-${end}/${size}`,
+      'Accept-Ranges': 'bytes',
+      'Content-Length': end - start + 1,
+      'Content-Type': mime.getType(path.extname(absoluteFilePath))!,
+    });
+
+    const readable = fs.createReadStream(absoluteFilePath, { start, end });
```
Comment (dgozman) [inline packages/playwright-core/src/utils/httpServer.ts:142]: Let's comment that both "range header and createReadStream accept inclusive start/end".

File: packages/playwright-core/src/utils/httpServer.ts [hunk 4]
```diff
@@ -80,22 +80,77 @@ export class HttpServer {
...
+    response.setHeader('Content-Type', contentType);
+    response.setHeader('Content-Length', content.byteLength);
+    response.end(content);
+  }
+
+  _serveRangeFile(request: http.IncomingMessage, response: http.ServerResponse, absoluteFilePath: string) {
+    const range = request.headers.range;
+    if (!range || !range.startsWith('bytes=') || range.includes(', ')) {
+      response.statusCode = 400;
+      return response.end('Bad request');
```
Comment (dgozman) [inline packages/playwright-core/src/utils/httpServer.ts:110]: Shouldn't this be `response.statusMessage = 'Bad request'`?

File: packages/playwright-core/src/utils/httpServer.ts (hunk 1 above)
Comment (dgozman) [reply packages/playwright-core/src/utils/httpServer.ts:114]: There could also be multiple dashes, like '1-5-10'. Let's error out in this case as well.

File: packages/playwright-core/src/utils/httpServer.ts (hunk 4 above)
Comment (mxschmitt) [reply packages/playwright-core/src/utils/httpServer.ts:110]: `statusMessage` is set automatically, having it in the response body is also common practise afaik.

File: packages/playwright-core/src/utils/httpServer.ts [hunk 5]
```diff
@@ -80,22 +80,77 @@ export class HttpServer {
...
+    response.statusCode = 200;
+    const contentType = mime.getType(path.extname(absoluteFilePath)) || 'application/octet-stream';
+    response.setHeader('Content-Type', contentType);
+    response.setHeader('Content-Length', content.byteLength);
+    response.end(content);
+  }
+
+  _serveRangeFile(request: http.IncomingMessage, response: http.ServerResponse, absoluteFilePath: string) {
+    const range = request.headers.range;
+    if (!range || !range.startsWith('bytes=') || range.includes(', ') || [...range].filter(char => char === '-').length !== 1) {
```
Comment (dgozman) [inline packages/playwright-core/src/utils/httpServer.ts:108]: nit: I'd do something like `match = range.match(/^bytes=(\d*)-(\d*)$/)` and use `+match[1], +match[2]`, but this also works.

**Review by dgozman — APPROVED** [review]

**Pull request merged**